### PR TITLE
LPS-186909: Add validation for API Application Entries if base URL value is valid or not

### PR DESCRIPTION
**Pull request by sergiojimcos**
**What is new?**
- Create `HeadlessBuilderValidator` to implement diferent validations for creating different entries using the API Builder. In this case is for the Base URL of an API Application
- Create `APIApplicationModelListener` which is going to check if the object definition is the API Application one and check if the base url is valid or not.
- Create `InvalidBaseURLException`.
- Add test cases for that.

**Description of the validation**

- When the user create or update an application, we have to validate that the Path only contains numbers, letters or "-". Blanks or special characters are not allowed.
- We need to return a specific error showing the valid character we allowe to be used.

**How to test it?**
- Deploy `headless-builder-api` and `headless-builder-impl`

**Related JIRA Ticket**
https://issues.liferay.com/browse/LPS-186909

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/validator/HeadlessBuilderValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * @author Sergio Jiménez del Coso
+ */
+@Component(service = HeadlessBuilderValidator.class)
+public class HeadlessBuilderValidatorImpl implements HeadlessBuilderValidator {
+
+	public boolean validatePath(String path) throws Exception {
+		Matcher matcher = _baseURLPattern.matcher(path);
+
+		return matcher.find();
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/validator/HeadlessBuilderValidatorImpl.java:31]: Should the validator throws the exception?

File: modules/apps/headless/headless-builder/headless-builder-api/src/main/java/com/liferay/headless/builder/exception/InvalidBaseURLException.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class InvalidBaseURLException extends BadRequestException {
+
+	public InvalidBaseURLException() {
+	}
+
+	public InvalidBaseURLException(String msg) {
+		super(msg);
+	}
+
+	public InvalidBaseURLException(String msg, Throwable throwable) {
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-api/src/main/java/com/liferay/headless/builder/exception/InvalidBaseURLException.java:31]: I think we don't need these methods

File: modules/apps/headless/headless-builder/headless-builder-api/src/main/java/com/liferay/headless/builder/exception/InvalidBaseURLException.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package com.liferay.headless.builder.exception;
+
+import javax.ws.rs.BadRequestException;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+public class InvalidBaseURLException extends BadRequestException {
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-api/src/main/java/com/liferay/headless/builder/exception/InvalidBaseURLException.java:22]: Do we need an extension of a `BadRequestException`? that is a JAX-RS exception. It is weird then that we need an ExceptionMapper

File: modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/jaxrs/exception/mapper/APIApplicationExceptionMapper.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package com.liferay.headless.builder.internal.jaxrs.exception.mapper;
+
+import com.liferay.headless.builder.exception.InvalidBaseURLException;
+import com.liferay.portal.vulcan.jaxrs.exception.mapper.BaseExceptionMapper;
+import com.liferay.portal.vulcan.jaxrs.exception.mapper.Problem;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+public class APIApplicationExceptionMapper
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/jaxrs/exception/mapper/APIApplicationExceptionMapper.java:24]: This name does not match the Liferay Pattern for exception mappers

File: modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/test/util/APIApplicationTestUtil.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.Serializable;
+
+import java.util.Map;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+public class APIApplicationTestUtil {
+
+	public static ObjectEntry addAPIApplicationEntry(
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/test/util/APIApplicationTestUtil.java:34]: I think you are only creating an ObjectEntry, so maybe we should rename the method and the name to `ObjectEntryTestUtil.addObjectEntry`

File: modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/model/listener/APIApplicationModelListener.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+@Component(service = ModelListener.class)
+public class APIApplicationModelListener
+	extends BaseModelListener<ObjectEntry> {
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/model/listener/APIApplicationModelListener.java:40]: I am not sure about the idea of creating a ModelListener. That means that this code will be executed any time an ObjectEntry is created or updated. That means for any kind of Object Entry

File: modules/apps/headless/headless-builder/headless-builder-api/src/main/java/com/liferay/headless/builder/exception/InvalidBaseURLException.java (hunk 3 above)
Comment (4lejandrito) [reply modules/apps/headless/headless-builder/headless-builder-api/src/main/java/com/liferay/headless/builder/exception/InvalidBaseURLException.java]: @sergiojimcos will try to remove it.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]
As discussed on the call:
* Rename the model listener to *ObjectEntryModelListener.
* Test at REST API level.
* Try to remove the exception mapper if possible.

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by 4lejandrito** [comment]
@sergiojimcos see https://github.com/liferay-headless/liferay-portal/pull/1289/commits/ad576a43b6e341295ad4faaa648b3300465a1aa5 and https://github.com/liferay-headless/liferay-portal/pull/1289/commits/637844d2d5ab37998122fcb3b754fa592a1e31a4.

**Comment by 4lejandrito** [comment]
ci:forward

**Review by 4lejandrito — APPROVED** [review]

**Comment by 4lejandrito** [comment]
@sergiojimcos we have to ignore the tests for now (@Ignore at class level). Because the object definition is not created in CI. Do it and ci:forward

**Comment by sergiojimcos** [comment]
ci:forward

**Comment by sergiojimcos** [comment]
Sent to Brian: https://github.com/brianchandotcom/liferay-portal/pull/136409

**Pull request closed without merging**